File: logwriter/cpp/Sample.cpp
/**
 * @file  Sample.cpp
 * @brief Usage sample for VizexecLogWriter
 *
 * @author Sunagae
 * @date 2011-02-11
 */

// Define ENABLE_VIZEXEC to enable VizexecLogWriter
// When not, VZE_*** macros will be ignored.
#define ENABLE_VIZEXEC

#include <string>
#include <iostream>
#include <cstdlib>
#include <boost/thread.hpp>
#include "VizexecLogWriter.hpp"

using namespace std;
using namespace boost;

const char *gMessageBuf = NULL;
const char *gAckBuf = NULL;
thread gProcessThread;

#define ms_sleep(t)     usleep((t) * 1000)

//------------------------------------------------------------------------------
// Functions for ProcessThread
//------------------------------------------------------------------------------
const char *WaitForMessage()
{
    VZE_FUNC
    while(!gMessageBuf)
        ms_sleep(1000);
    VZE_RECV1(gMessageBuf)
    return gMessageBuf;
}

const char *ProcessMessage(const char *msg)
{
    VZE_FUNC
    ms_sleep(1000);
    VZE_EVENT("MessageProcessEvent")
    // do something
    ms_sleep(2000);
    return "<ackmessage>";
}

void SendAck(const char *msg)
{
    VZE_FUNC
    VZE_SEND1(msg)
    gAckBuf = msg;
}

void process_thread()
{
    VZE_FUNCCUSTOM("ThreadMain")
    VZE_THREADNAME("ProcessThread")
    while(true)
    {
        const char *msg = WaitForMessage();
        VZE_PHASE("Process phase")
        const char *result = ProcessMessage(msg);
        SendAck(result);
    }
}

//------------------------------------------------------------------------------
// Functions for MainThread
//------------------------------------------------------------------------------
void CreateThread()
{
    VZE_FUNC
    gProcessThread = thread(&process_thread);
    ms_sleep(1000);
}


void SendMessage(const char *msg)
{
    VZE_FUNC
    VZE_SEND1(msg)
    gMessageBuf = msg;
}

const char* WaitForAck()
{
    VZE_FUNC
    while(!gAckBuf)
        ms_sleep(1000);
    VZE_RECV1(gAckBuf)
    return gAckBuf;    
}

void ShowResult(const char *result)
{
    VZE_FUNC
    cout << "Result: " << result << endl;
    VZE_EVENT("DrawEvent")
}

int main()
{
    VZE_START
    VZE_THREADNAME("MainThread")
    VZE_FUNC
    CreateThread();
    SendMessage("testmessage");
    const char *result = WaitForAck();
    ShowResult(result);

}

File: logwriter/python/sample.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import VizexecLogWriter as vze

gMessageBuf = None
gAckBuf = None
#------------------------------------------------------------------------------
# Functions for ProcessThread
#------------------------------------------------------------------------------
@vze.func
def WaitForMessage():
    while not gMessageBuf:
        time.sleep(1.0)
    vze.recv(gMessageBuf)
    return gMessageBuf

@vze.func
def ProcessMessage(msg):
    time.sleep(1000)
    vze.event("MessageProcessEvent")
    # do something
    time.sleep(2.0)
    return "<ackmessage>"

@vze.func
def SendAck(msg):
    global gAckBuf
    VZE_SEND1(msg)
    gAckBuf = msg

@vze.func_custom("ThreadMain")
def process_thread():
    vze.thread_name("ProcessThread")
    while True:
        msg = WaitForMessage()
        vze.phase("Process phase")
        result = ProcessMessage(msg)
        SendAck(result)

#------------------------------------------------------------------------------
# Functions for MainThread
#------------------------------------------------------------------------------
@vze.func
def CreateThread():
    t = threading.Thread(target = process_thread)
    t.setDaemon(True)
    t.start()
    time.sleep(1000)


@vze.func
def SendMessage(msg):
    global gMessageBuf
    vze.send(msg)
    gMessageBuf = msg

@vze.func
def WaitForAck():
    global gAckBuf
    while not gAckBuf:
        time.sleep(1000)
    vze.recv(gAckBuf)
    return gAckBuf

@vze.func
def ShowResult(result):
    print "Result: ", result;
    vze.event("DrawEvent")

@vze.func
def main():
    vze.thread_name("MainThread")
    CreateThread()
    SendMessage("testmessage")
    result = WaitForAck()
    ShowResult(result)

vze.start()
main()


